src/libraries/UniswapV3Lib.sol

```diff
@@ -139,6 +222,132 @@ library UniswapV3Lib {
+                token0         : cache.token0,
+                token1         : cache.token1,
+                fee            : cache.fee,
+                tickLower      : params.tick.lower,
+                tickUpper      : params.tick.upper,
+                recipient      : address(context.proxy),
+                amount0Desired : params.amountDesired.amount0,
+                amount1Desired : params.amountDesired.amount1,
+                amount0Min     : params.amountMin.amount0,
+                amount1Min     : params.amountMin.amount1,
+                deadline       : context.deadline
+            });
+
+        bytes memory result = context.proxy.doCall(
+            address(params.positionManager),
+            abi.encodeCall(
+                INonfungiblePositionManager.mint,
+                (mintParams)
+            )
+        );
+
+        (tokenId, liquidity, amount0, amount1) = abi.decode(result, (uint256, uint128, uint256, uint256));
+    }
+
+    function _addLiquidityToExistingPosition(UniV3Context calldata context, AddLiquidityParams calldata params)
+        internal
+        returns (uint128 liquidity, uint256 amount0, uint256 amount1)
+    {
+        INonfungiblePositionManager.IncreaseLiquidityParams memory increaseLiquidityParams
+            = INonfungiblePositionManager.IncreaseLiquidityParams({
+                tokenId        : params.tokenId,
+                amount0Desired : params.amountDesired.amount0,
+                amount1Desired : params.amountDesired.amount1,
+                amount0Min     : params.amountMin.amount0,
+                amount1Min     : params.amountMin.amount1,
+                deadline       : context.deadline
+            });
+
+        bytes memory result = context.proxy.doCall(
+            address(params.positionManager),
+            abi.encodeCall(
+                INonfungiblePositionManager.increaseLiquidity,
+                (increaseLiquidityParams)
+            )
+        );
+
+        (liquidity, amount0, amount1) = abi.decode(result, (uint128, uint256, uint256));
+    }
+
+    function _validateAddLiquidityMinAmounts(UniV3Context calldata context, AddLiquidityParams calldata params) internal view {
+        // Fetch twap tick
+        (int24 twapTick, ) = UniswapV3OracleLib.consult(context.pool, params.twapSecondsAgo);
+
+        uint160 sqrtTwapPriceX96   = TickMath.getSqrtRatioAtTick(twapTick);
+        uint160 sqrtRatioLowerX96  = TickMath.getSqrtRatioAtTick(params.tick.lower);
+        uint160 sqrtRatioUpperX96  = TickMath.getSqrtRatioAtTick(params.tick.upper);
+
+        uint128 expectedLiquidity = LiquidityAmounts.getLiquidityForAmounts(
+            sqrtTwapPriceX96,
+            sqrtRatioLowerX96,
+            sqrtRatioUpperX96,
+            params.amountDesired.amount0,
+            params.amountDesired.amount1
+        );
+
+        uint256 expectedAmount0;
+        uint256 expectedAmount1;
+
+        if (expectedLiquidity > 0) {
```

Comment:
A little bit of a hard time reasoning about `expectedLiquidity = 0` in this case. Are we sure that we want to continue? It seems like the min threshold requires would then pass.